### PR TITLE
intel-oneapi-mkl: linking with the +cluster variant is broken for external mkl

### LINKED ISSUE #38238: intel-oneapi-mkl: linking with the +cluster variant is broken for external mkl
### Steps to reproduce

1. Spec an external installation of MKL as `intel-oneapi-mkl +cluster` in `packages.yaml`
2. Install any package with the above `intel-oneapi-mkl +cluster` as a dependency.

This bug was introduced in https://github.com/spack/spack/pull/37386. Since externals don't have dependencies in their specs, the reference to mpi fails and results in an error.

It would be ideal to have a fix which accomplishes the same goal as the above PR but in a way which works for external MKL as well, as the PR is valuable for fixing build errors which result when building serial libraries, for example `blaspp ^intel-oneapi-mkl +cluster`.

### Error message

``` console
==> Error: KeyError: 'No spec with name mpi in intel-oneapi-mkl@2023.1.0%oneapi@2023.1.0+cluster~envmods~ilp64+shared build_system=generic arch=linux-sles15-x86_64'

.../spack/var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py:168, in _find_mkl_libs:
        165
        166        resolved_libs = find_libraries(libs, lib_path, shared=shared)
        167        if "+cluster" in self.spec:
  >>    168            resolved_libs = resolved_libs + self.spec["mpi"].libs
        169        return resolved_libs
```

### Information on your system

* **Spack:** 0.20.0.dev0 (b1b7dab180ba0be240dd8983d8173fa1590ef080)
* **Python:** 3.6.15
* **Platform:** linux-sles15-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by wrtobin**
fixes #38238.

### CODE REVIEW

**Review by rscohn2 — DISMISSED** [review]
Thanks for fixing this issue!

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 1]
```diff
@@ -182,19 +191,13 @@ def _find_mkl_libs(self, shared):
         libs.append("libmkl_core")
 
         if self.spec.satisfies("+cluster"):
-            if any(
-                self.spec.satisfies(m)
-                for m in ["^intel-oneapi-mpi", "^intel-mpi", "^mpich", "^cray-mpich"]
-            ):
+            if self.spec.satisfies("mpi=mpich"):
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py]: After some more thought:
this change will break the package for some existing users. Could you instead leave line 187 and add "mpi=mpich" to the list there? Same for "mpi=openmpi" below

**Review by rscohn2 — DISMISSED** [review]

**Review by rscohn2 — DISMISSED** [review]

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 2]
```diff
@@ -103,6 +103,15 @@ class IntelOneapiMkl(IntelOneApiLibraryPackage):
...
         "cluster", default=False, description="Build with cluster support: scalapack, blacs, etc"
     )
+    variant(
+        "mpi",
+        default="none",
+        values=("none", "mpich", "openmpi"),
+        description="MPI family",
+        multi=False,
+    )
+    conflicts("mpi=none", when="+cluster")
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py]:
```suggestion
```

To preserve old behavior without mpi=x

**Comment by rscohn2** [comment]
I don't understand the gitlab failure. Maybe mpi is a reserved keyword? You could change the variant name to "mpi_family" or some other name.

**Comment by rscohn2** [comment]
I verified locally that renaming the variant to something other than mpi makes the gitlab issue go away.

**Comment by wrtobin** [comment]
Thanks for verifying

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 3]
```diff
@@ -184,16 +192,22 @@ def _find_mkl_libs(self, shared):
...
-            elif self.spec.satisfies("^openmpi"):
+            elif any(self.spec.satisfies(m) for m in ["^openmpi", "mpi_family=openmpi"]):
                 libs.append(self._xlp64_lib("libmkl_blacs_openmpi"))
             else:
                 raise RuntimeError(
                     (
-                        "intel-oneapi-mpi +cluster requires one of "
-                        "^intel-oneapi-mpi, ^intel-mpi, ^mpich, or ^openmpi"
+                        "intel-oneapi-mpi +cluster requires one of ^intel-oneapi-mpi, "
+                        "^intel-mpi, ^mpich, ^cray-mpich, mpi=mpich, ^openmpi, or mpi=openmpi"
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py]:
```suggestion
                        "^intel-mpi, ^mpich, ^cray-mpich, mpi_family=mpich, ^openmpi, or mpi_family=openmpi"
```

**Review by rscohn2 — DISMISSED** [review]

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 4]
```diff
@@ -184,16 +192,22 @@ def _find_mkl_libs(self, shared):
...
-            elif self.spec.satisfies("^openmpi"):
+            elif any(self.spec.satisfies(m) for m in ["^openmpi", "mpi_family=openmpi"]):
                 libs.append(self._xlp64_lib("libmkl_blacs_openmpi"))
             else:
                 raise RuntimeError(
                     (
-                        "intel-oneapi-mpi +cluster requires one of "
-                        "^intel-oneapi-mpi, ^intel-mpi, ^mpich, or ^openmpi"
+                        "intel-oneapi-mpi +cluster requires one of ^intel-oneapi-mpi, "
+                        "^intel-mpi, ^mpich, ^cray-mpich, mpi_family=mpich, ^openmpi, or mpi_family=openmpi"
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py]:
```suggestion
                        "^intel-mpi, ^mpich, ^cray-mpich, mpi_family=mpich, ^openmpi, "
                        "or mpi_family=openmpi"
```

**Review by rscohn2 — APPROVED** [review]

**Pull request merged**